src/modm/platform/dac/sam_x7x/dac_impl.hpp.in

```
@@ -0,0 +1,190 @@
+
+		Ch1Pin::setInput();
+	}
+%% if target.pin != "j"
+	if constexpr (!std::is_void_v<DatrgPin>) {
+		using DatrgConnector = typename DatrgPin::template Connector<Peripheral, Peripheral::Datrg>;
+		DatrgConnector::connect();
+	}
+%% endif
+}
+
+template<typename SystemClock, frequency_t frequency, percent_t tolerance>
+inline void
+Dac::initialize()
+{
+	ClockGen::enable<ClockPeripheral::Dacc>();
+
+	DACC->DACC_CR = DACC_CR_SWRST;
+
+	constexpr auto result = modm::Prescaler::from_function(
+		SystemClock::Dacc, frequency, 0, 15,
+		[](uint32_t x) { return (x + 2); }
+	);
+
+	static_assert(result.frequency <= MaxClock,
+					"Generated DAC clock is above maximum frequency!");
+	assertBaudrateInTolerance<result.frequency, frequency, tolerance>();
+
+	DACC->DACC_MR = DACC_MR_PRESCALER(result.index);
+
+	// set analog bias, see datasheet
+	if constexpr (result.frequency / 12.f > kHz(500)) {
+		DACC->DACC_ACR = DACC_ACR_IBCTLCH1(3) | DACC_ACR_IBCTLCH0(3);
+	} else {
+		DACC->DACC_ACR = DACC_ACR_IBCTLCH1(1) | DACC_ACR_IBCTLCH0(1);
+	}
+}
+
+inline void
+Dac::enableChannel(Channel chan)
+{
+	DACC->DACC_CHER = 1u << static_cast<uint32_t>(chan);
+}
+
+inline void
+Dac::disableChannel(Channel chan)
+{
+	DACC->DACC_CHDR = 1u << static_cast<uint32_t>(chan);
+}
+
+inline void
+Dac::setChannelMode(Channel chan, Mode mode)
+{
+	if (mode == Mode::FreeRunning) {
+		// disable trigger mode
+		DACC->DACC_TRIGR &= ~(DACC_TRIGR_TRGEN0_EN << uint32_t(chan));
+		// disable max speed mode
+		DACC->DACC_MR &= ~(DACC_MR_MAXS0_Msk << uint32_t(chan));
+	} else if (mode == Mode::Trigger) {
+		// enable trigger mode
+		DACC->DACC_TRIGR |= (DACC_TRIGR_TRGEN0_EN << uint32_t(chan));
+		// disable max speed mode
+		DACC->DACC_MR &= ~(DACC_MR_MAXS0_Msk << uint32_t(chan));
+	} else {
+		// enable max speed mode
+		DACC->DACC_MR |= (DACC_MR_MAXS0_Msk << uint32_t(chan));
+	}
+}
+
+inline void
+Dac::setOutput(Channel chan, uint32_t value)
+{
+	DACC->DACC_CDR[uint32_t(chan)] = value;
+}
```

Fast, but dangerous. Ok for me 🤷🏽

---

Maybe we could prevent errors with this:
```suggestion
inline void
Dac::setOutput(Channel chan, uint32_t value)
{
	DACC->DACC_CDR[uint32_t(chan) & 0b1] = value;
}
```

---

Hm, not sure about that one. It's pretty hard to get that wrong unintentionally with the `enum class` parameter. You have to force it with a cast:
```
auto channel = static_cast<Channel>(5);
…
Dac::setOutput(channel, 0);
```
If I really made that mistake I'd rather want my code not to work normally. That could mask errors in other places.

If you prefer the change with the `& 1` we could add that, I am not strongly against it.

---

Ahh yeah, I somehow thought `chan` is an `uint32_t`.